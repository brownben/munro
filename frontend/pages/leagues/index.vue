<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
