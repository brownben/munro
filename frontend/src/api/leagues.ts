import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
