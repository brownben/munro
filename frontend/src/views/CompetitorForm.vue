<!--
  Competitor Form

  The form for Creating/ Updating Competitor
-->

<template>
  <div class="view">
    <vue-headful
      v-if="create"
      title="Munro - Create Competitor"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <vue-headful
      v-else
      title="Munro - Edit Competitor"
      description
      :head="{
        'meta': {name: 'robots', content:'noindex'},
      }"
    />
    <div v-if="!notFound">
      <h1 v-if="create" class="text-main text-3xl font-normal font-heading mb-2">Create Competitor</h1>
      <h1 v-if="!create" class="text-main text-3xl font-normal font-heading mb-2">Edit Competitor</h1>
      <!-- @submit on submit via enter key in the last field, .prevent prevents page reload -->
      <form @submit.prevent="submit">
        <text-input v-model.trim="name" label="Name:" />
        <text-input v-model.trim="club" label="Club:" />
        <text-input v-model.trim="ageClass" label="Age Class:" />
        <dropdown-input
          v-model="league"
          :list="leagues.map(league => league.name)"
          label="League:"
        />
        <dropdown-input v-model="course" :list="coursesInLeague" label="Course:" />
        <button v-if="create" class="button-lg">Create Competitor</button>
        <button v-if="!create" class="button-lg">Update Competitor</button>
      </form>
    </div>
    <not-found v-if="notFound" />
  </div>
</template>

<script>
import axios from 'axios'
import DropdownInput from '@/components/DropdownInput'
import TextInput from '@/components/TextInput'
const NotFound = () => import('@/views/NotFound')

export default {
  components: {
    'NotFound': NotFound,
    'DropdownInput': DropdownInput,
    'TextInput': TextInput,
  },

  data: function () {
    return {
      id: '',
      notFound: false,
      create: true,
      leagues: [],
      name: '',
      club: '',
      ageClass: '',
      league: '',
      course: '',
    }
  },

  computed: {
    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(league => league.name === this.league)
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },
  },

  // On Load
  mounted: function () {
    this.getLeagues()
    if (this.$route.path.includes('edit')) {
      this.create = false
      this.getCompetitorDetails()
    }
  },

  methods: {
    submit: function () {
      if (this.create) this.createCompetitor()
      else this.updateCompetitor()
    },

    getCompetitorDetails: function () {
      return axios.get('/api/competitors/' + this.$route.params.id)
        .then(response => {
          if (!response.data) this.notFound = true
          else {
            this.id = this.$route.params.id
            this.name = response.data.name
            this.club = response.data.club
            this.ageClass = response.data.ageClass
            this.league = response.data.league
            this.course = response.data.course
          }
        })
        .catch(() => this.$messages.addMessage('Problem Getting Competitor Details'))
    },

    getLeagues: function () {
      return axios.get('/api/leagues')
        .then(response => { this.leagues = response.data })
        .catch(() => this.$messages.addMessage('Problem Fetching List of Leagues'))
    },

    validateForm: function () {
      return (this.name !== '' && this.league !== '' && this.course !== '')
    },

    createCompetitor: function () {
      if (this.validateForm()) {
        return axios.post('/api/competitors', {
          name: this.name,
          club: this.club,
          ageClass: this.ageClass,
          league: this.league,
          course: this.course,
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(() => this.$messages.addMessage('Error: Problem Creating Competitor - Please Try Again'))
      }
      else this.$messages.addMessage('Please Ensure Name and League Fields are not Blank')
    },

    updateCompetitor: function () {
      if (this.validateForm()) {
        return axios.put('/api/competitors/' + this.id, {
          id: this.id,
          name: this.name,
          club: this.club,
          ageClass: this.ageClass,
          league: this.league,
          course: this.course,
        })
          .then(response => this.returnToCompetitorsPage(response))
          .catch(error => this.$messages.addMessage(error.response.data.message))
      }
      else this.$messages.addMessage('Please Ensure Name and League Fields are not Blank')
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$messages.addMessage(response.data.message)
      this.$router.push('/competitors')
    },
  },
}

</script>
