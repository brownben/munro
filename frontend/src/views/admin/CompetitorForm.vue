<!--
  Competitor Form

  The form for Creating/ Updating Competitor
-->

<template>
  <div>
    <Meta
      :head="{
        meta: { name: 'robots', content: 'noindex' },
      }"
      :title="`Munro - ${title}`"
      description
    />

    <Layout v-if="!notFound" :title="title">
      <!-- @submit on submit via enter key in the last field, .prevent prevents page reload -->
      <form class="col-span-2" @submit.prevent="submit">
        <TextInput v-model.trim="name" label="Name:" />
        <TextInput v-model.trim="club" label="Club:" class="mt-4" />
        <TextInput v-model.trim="ageClass" label="Age Class:" class="mt-4" />
        <DropdownInput
          v-model="league"
          :list="leagues.map((league) => league.name)"
          label="League:"
          class="mt-4"
        />
        <DropdownInput
          v-model="course"
          :list="coursesInLeague"
          label="Course:"
          class="mt-4"
        />

        <button v-if="create" class="mt-8 button-lg">Create Competitor</button>
        <button v-else class="mt-8 button-lg">Update Competitor</button>
      </form>
    </Layout>
    <not-found v-else />
  </div>
</template>

<script>
import { defineAsyncComponent } from 'vue'
import axios from 'axios'

import Layout from '/@/components/Layout.vue'
import DropdownInput from '/@/components/inputs/DropdownInput.vue'
import TextInput from '/@/components/inputs/TextInput.vue'

const NotFound = defineAsyncComponent(() => import('/@/views/NotFound.vue'))

export default {
  components: {
    Layout,
    DropdownInput,
    TextInput,
    NotFound,
  },

  data: function () {
    return {
      id: '',
      notFound: false,
      create: true,
      leagues: [],
      name: '',
      club: '',
      ageClass: '',
      league: '',
      course: '',
    }
  },

  computed: {
    title: function () {
      if (this.create) return 'Create Competitor'
      else return 'Edit Competitor'
    },

    coursesInLeague: function () {
      const selectedLeague = this.leagues.filter(
        (league) => league.name === this.league
      )
      if (selectedLeague.length > 0) return selectedLeague[0].courses
      else return []
    },
  },

  // On Load
  mounted: function () {
    this.getLeagues()
    if (this.$route.path.includes('edit')) {
      this.create = false
      this.getCompetitorDetails()
    }
  },

  methods: {
    submit: function () {
      if (this.create) this.createCompetitor()
      else this.updateCompetitor()
    },

    getCompetitorDetails: function () {
      return axios
        .get(`/api/competitors/${this.$route.params.id}`)
        .then((response) => {
          if (!response.data) this.notFound = true
          else {
            this.id = this.$route.params.id
            this.name = response.data.name
            this.club = response.data.club
            this.ageClass = response.data.ageClass
            this.league = response.data.league
            this.course = response.data.course
          }
        })
        .catch(() =>
          this.$store.dispatch(
            'createMessage',
            'Problem Getting Competitor Details'
          )
        )
    },

    getLeagues: function () {
      return axios
        .get('/api/leagues')
        .then((response) => {
          this.leagues = response.data
        })
        .catch(() =>
          this.$store.dispatch(
            'createMessage',
            'Problem Fetching List of Leagues'
          )
        )
    },

    validateForm: function () {
      return this.name !== '' && this.league !== '' && this.course !== ''
    },

    createCompetitor: function () {
      if (this.validateForm()) {
        return axios
          .post('/api/competitors', {
            name: this.name,
            club: this.club,
            ageClass: this.ageClass,
            league: this.league,
            course: this.course,
          })
          .then((response) => this.returnToCompetitorsPage(response))
          .catch(() =>
            this.$store.dispatch(
              'createMessage',
              'Error: Problem Creating Competitor - Please Try Again'
            )
          )
      } else
        this.$store.dispatch(
          'createMessage',
          'Please Ensure Name and League Fields are not Blank'
        )
    },

    updateCompetitor: function () {
      if (this.validateForm()) {
        return axios
          .put(`/api/competitors/${this.id}`, {
            id: this.id,
            name: this.name,
            club: this.club,
            ageClass: this.ageClass,
            league: this.league,
            course: this.course,
          })
          .then((response) => this.returnToCompetitorsPage(response))
          .catch((error) =>
            this.$store.dispatch('createMessage', error.response.data.message)
          )
      } else
        this.$store.dispatch(
          'createMessage',
          'Please Ensure Name and League Fields are not Blank'
        )
    },

    returnToCompetitorsPage: function (response) {
      // Go to league page after successful update/ creation
      this.$store.dispatch('createMessage', response.data.message)
      this.$router.push(`/leagues/${this.league}/competitors`)
    },
  },
}
</script>
