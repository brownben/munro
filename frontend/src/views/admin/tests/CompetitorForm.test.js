import axios from 'axios'
import { mount, shallowMount } from '@vue/test-utils'
import CompetitorForm from '@/views/admin/CompetitorForm.vue'
import { sampleSingleLeague } from '@/tests/test data/leagues'

jest.mock('axios')

beforeEach(() => {
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: [] })
})

test('Is a Vue Instance', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '' },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - Create', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeTruthy()
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - Upload', () => {
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.vm.create).toBeFalsy()
  expect(wrapper.element).toMatchSnapshot()
})

test('Return from Page', () => {
  const mockRouterPush = jest.fn()
  const mockAddMessage = jest.fn()
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: mockAddMessage },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
      $router: { push: mockRouterPush },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: 'a' })
  wrapper.vm.returnToCompetitorsPage({ data: { message: 'Hello' } })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Hello')
  expect(mockRouterPush).toHaveBeenCalledTimes(1)
  expect(mockRouterPush).toHaveBeenLastCalledWith('/leagues/a/competitors')
})

test('Get Leagues - Request Called Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '', params: { id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: sampleSingleLeague })
  await wrapper.vm.getLeagues()
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues')
})

test('Get Leagues - Processes Response Correctly', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '', params: { id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockResolvedValue({ data: [{ league: '1' }, { league: '2' }] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([{ league: '1' }, { league: '2' }])

  axios.get.mockResolvedValue({ data: [] })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])

  axios.get.mockResolvedValue({
    data: [
      { league: 'a', property: 3 },
      { league: '2', another: 4 },
    ],
  })
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([
    { league: 'a', property: 3 },
    { league: '2', another: 4 },
  ])
})

test('Get Leagues - Shows Message on Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '', params: { id: '' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeagues()
  expect(wrapper.vm.leagues).toEqual([])
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Fetching List of Leagues'
  )
})

test('Courses in League', () => {
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: 'a', leagues: [] })
  expect(wrapper.vm.coursesInLeague).toEqual([])
  wrapper.setData({
    leagues: [
      { name: 'a', courses: [2] },
      { name: 'b', courses: 3 },
      { name: 'c', courses: [4] },
    ],
  })
  expect(wrapper.vm.coursesInLeague).toEqual([2])
  wrapper.setData({
    league: 'c',
    leagues: [
      { name: 'a', courses: 2 },
      { name: 'b', courses: 3 },
      { name: 'c', courses: [4] },
    ],
  })
  expect(wrapper.vm.coursesInLeague).toEqual([4])
})

test('Validate', () => {
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ name: '', league: '', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: 'a', league: '', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: '', league: 'a', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: '', league: '', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: 'a', league: 'a', course: '' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: 'a', league: '', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: '', league: 'a', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeFalsy()
  wrapper.setData({ name: 'a', league: 'a', course: 'a' })
  expect(wrapper.vm.validateForm()).toBeTruthy()
})

test('Submit', async () => {
  const mockCreate = jest.fn()
  const mockUpdate = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: jest.fn() },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setMethods({
    createCompetitor: mockCreate,
    updateCompetitor: mockUpdate,
  })
  wrapper.setData({ create: true })
  await wrapper.vm.submit()
  expect(mockUpdate).toHaveBeenCalledTimes(0)
  expect(mockCreate).toHaveBeenCalledTimes(1)
  wrapper.setData({ create: false })
  await wrapper.vm.submit()
  expect(mockCreate).toHaveBeenCalledTimes(1)
  expect(mockUpdate).toHaveBeenCalledTimes(1)
})

test('Create Competitor - Invalid', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await wrapper.vm.createCompetitor()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )
})

test('Create Competitor - Correct API', async () => {
  const wrapper = shallowMount(CompetitorForm, {
    stubs: ['router-link', 'vue-headful'],
    mocks: {
      $messages: { addMessage: jest.fn() },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
  })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.post.mockRejectedValue({ response: { data: { message: 'Hello' } } })
  await wrapper.vm.createCompetitor()
  expect(axios.post).toHaveBeenCalledTimes(1)
  expect(axios.post).toHaveBeenLastCalledWith(
    '/api/competitors',
    expect.any(Object)
  )
})

test('Create Competitor - Success', async () => {
  const mockReturnToCompetitorsPage = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    stubs: ['router-link', 'vue-headful'],
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
  })
  wrapper.setMethods({ returnToCompetitorsPage: mockReturnToCompetitorsPage })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.post.mockResolvedValue({ data: { message: 'Hello' } })
  await wrapper.vm.createCompetitor()
  expect(mockReturnToCompetitorsPage).toHaveBeenCalledTimes(1)
})

test('Create Competitor - Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.post.mockRejectedValue({ response: { data: { message: 'Hello' } } })
  await wrapper.vm.createCompetitor()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Error: Problem Creating Competitor - Please Try Again'
  )
})

test('Update Competitor - Invalid', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await wrapper.vm.updateCompetitor()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Please Ensure Name and League Fields are not Blank'
  )
})

test('Update Competitor - Correct API', async () => {
  const wrapper = shallowMount(CompetitorForm, {
    stubs: ['router-link', 'vue-headful'],
    mocks: {
      $messages: { addMessage: jest.fn() },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
  })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.put.mockRejectedValue({ response: { data: { message: 'Hello' } } })
  await wrapper.vm.updateCompetitor()
  expect(axios.put).toHaveBeenCalledTimes(1)
  expect(axios.put).toHaveBeenLastCalledWith(
    '/api/competitors/',
    expect.any(Object)
  )
})

test('Update Competitor - Success', async () => {
  const mockReturnToCompetitorsPage = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    stubs: ['router-link', 'vue-headful'],
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
  })
  wrapper.setMethods({ returnToCompetitorsPage: mockReturnToCompetitorsPage })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.put.mockResolvedValue({ data: { message: 'Hello' } })
  await wrapper.vm.updateCompetitor()
  expect(mockReturnToCompetitorsPage).toHaveBeenCalledTimes(1)
})

test('Update Competitor - Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $messages: { addMessage: mockAddMessageFunction },
      $route: { path: '/competitors/1/edit', params: { id: '' } },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ name: 'a', league: 'b', course: 'c' })
  axios.put.mockRejectedValue({ response: { data: { message: 'Hello' } } })
  await wrapper.vm.updateCompetitor()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith('Hello')
})

test('Get Competitor Details - Error', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      created: () => {},
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitorDetails()
  expect(mockAddMessageFunction).toHaveBeenCalledTimes(1)
  expect(mockAddMessageFunction).toHaveBeenLastCalledWith(
    'Problem Getting Competitor Details'
  )
})

test('Get Competitor Details - Not Found', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = shallowMount(CompetitorForm, {
    mocks: {
      $route: { path: '/leagues/1/edit', params: { name: '' } },
      $messages: { addMessage: mockAddMessageFunction },
      created: () => {},
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: false })
  await wrapper.vm.getCompetitorDetails()
  expect(wrapper.vm.notFound).toBeTruthy()
  expect(wrapper.findAll('form').length).toBe(0)
})

test('Get Competitor Details - Correct API Call', async () => {
  const mockAddMessageFunction = jest.fn()
  const wrapper = mount(CompetitorForm, {
    mocks: {
      $route: { path: '/competitors/1/edit', params: { id: 'a' } },
      $messages: { addMessage: mockAddMessageFunction },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getCompetitorDetails()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/competitors/a')
})
