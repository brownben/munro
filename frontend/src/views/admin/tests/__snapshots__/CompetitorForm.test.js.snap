// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - Create 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Create Competitor"
  />
   
  <div
    class="w-10/12 mx-auto md:w-3/4 xl:w-2/3 md:mt-2"
  >
    <h1
      class="my-2 text-3xl font-bold md:text-35xl font-heading text-main-700"
    >
      
    Create Competitor
  
    </h1>
     
    <div
      class="grid grid-cols-2 gap-8 pb-8 md:gap-6 lg:gap-8"
    >
      <form
        class="col-span-2"
      >
        <div>
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Name:"
          >
            Name:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Name:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Club:"
          >
            Club:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Club:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Age Class:"
          >
            Age Class:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Age Class:"
            type="text"
          />
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="League:"
          >
            
    League:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="League:"
          >
            <option />
             
          </select>
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Course:"
          >
            
    Course:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Course:"
          >
            <option />
             
          </select>
        </div>
         
        <button
          class="mt-8 button-lg"
        >
          Create Competitor
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`Renders Correctly - Upload 1`] = `
<div>
  <vue-headful-stub
    description=""
    head="[object Object]"
    title="Munro - Edit Competitor"
  />
   
  <div
    class="w-10/12 mx-auto md:w-3/4 xl:w-2/3 md:mt-2"
  >
    <h1
      class="my-2 text-3xl font-bold md:text-35xl font-heading text-main-700"
    >
      
    Edit Competitor
  
    </h1>
     
    <div
      class="grid grid-cols-2 gap-8 pb-8 md:gap-6 lg:gap-8"
    >
      <form
        class="col-span-2"
      >
        <div>
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Name:"
          >
            Name:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Name:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Club:"
          >
            Club:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Club:"
            type="text"
          />
        </div>
         
        <div
          class="mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Age Class:"
          >
            Age Class:
          </label>
           
          <input
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Age Class:"
            type="text"
          />
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="League:"
          >
            
    League:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="League:"
          >
            <option />
             
          </select>
        </div>
         
        <div
          class="relative mt-4"
        >
          <label
            class="block pb-1 select-none font-heading text-main-600"
            for="Course:"
          >
            
    Course:
  
          </label>
           
          <svg
            class="absolute bottom-0 right-0 h-6 mr-2 text-main-500"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            style="margin-bottom: 0.625rem;"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 9l-7 7-7-7"
            />
          </svg>
           
          <select
            class="w-full px-3 py-2 transition-all duration-300 border-2 border-opacity-0 outline-none appearance-none rounded-shape text-body bg-main-100 border-main focus:border-opacity-100"
            id="Course:"
          >
            <option />
             
          </select>
        </div>
         
        <button
          class="mt-8 button-lg"
        >
          Update Competitor
        </button>
      </form>
    </div>
  </div>
</div>
`;
